service/storage_proxy.cc

```diff
@@ -498,22 +505,25 @@ class storage_proxy::remote {
             smp_service_group smp_grp, const rpc::client_info& cinfo, rpc::opt_time_point t,
             frozen_mutation in, inet_address_vector_replica_set forward, gms::inet_address reply_to,
             unsigned shard, storage_proxy::response_id_type response_id,
-            rpc::optional<std::optional<tracing::trace_info>> trace_info, rpc::optional<db::per_partition_rate_limit::info> rate_limit_info_opt) {
+            rpc::optional<std::optional<tracing::trace_info>> trace_info,
+            rpc::optional<db::per_partition_rate_limit::info> rate_limit_info_opt,
+            rpc::optional<fencing_token> fence) {
         tracing::trace_state_ptr trace_state_ptr;
         auto src_addr = netw::messaging_service::get_source(cinfo);
         auto rate_limit_info = rate_limit_info_opt.value_or(std::monostate());
 
         auto schema_version = in.schema_version();
         return handle_write(src_addr, t, schema_version, std::move(in), std::move(forward), reply_to, shard, response_id,
                 trace_info ? *trace_info : std::nullopt,
+                fence.value_or(fencing_token{}),
                 /* apply_fn */ [smp_grp, rate_limit_info] (shared_ptr<storage_proxy>& p, tracing::trace_state_ptr tr_state, schema_ptr s, const frozen_mutation& m,
-                        clock_type::time_point timeout) {
-                    return p->mutate_locally(std::move(s), m, std::move(tr_state), db::commitlog::force_sync::no, timeout, smp_grp, rate_limit_info);
+                        clock_type::time_point timeout, fencing_token fence) {
+                    return p->apply_fence(p->mutate_locally(std::move(s), m, std::move(tr_state), db::commitlog::force_sync::no, timeout, smp_grp, rate_limit_info), fence);
```

Comment:
I don't understand this. We apply the write to the replica - but then lie to the coordinator that it was fenced? What's the point if it wasn't really fenced, we actually wrote it?

Checking the version at any moment after the data was written doesn't make sense to me. And in the commit message you write:
> apply_fence after the local write guarantees that no mutations succeed on coordinators if the fence version has been updated on the replica during the write.

We may reply success, but then our fence version changes while the reply message is in transit to the coordinator. What then? To an external observer, this is equivalent to the fencing version changing "during the write". Basically "duration of the write" is between coordinator generating the write and getting a response. Not the duration of `mutate_locally` on the replica.

Reply:
> 
> We may reply success, but then our fence version changes while the reply message is in transit to the coordinator. What then? To an external observer, this is equivalent to the fencing version changing "during the write". Basically "duration of the write" is between coordinator generating the write and getting a response. Not the duration of `mutate_locally` on the replica.

If the version changes after reply it means it changed after the data was written and after the barrier done by the topology coordinator completed and hence it means that the data will be included in streaming and thus it is valid to reply with success to the write.

Reply:
Ok, in other words
- if the version did not increase after the write was done on the replica, then we can be sure that this write will be included in streaming. So we can respond with ACK
- if we see that the version increased after the write was done on the replica, then we're not sure if this write will be included in streaming. Therefore we shouldn't ACK even though we successfully persisted the write.

In the second case it doesn't matter that the version may increase while the ACK is in transit. The write will persist.

So it makes sense. Thanks.